src/matchengine/controller.rs

```diff
@@ -552,6 +553,64 @@ impl Controller {
         Ok(())
     }
 
+    pub fn transfer(&mut self, real: bool, req: TransferRequest) -> Result<TransferResponse, Status> {
+        if !self.check_service_available() {
+            return Err(Status::unavailable(""));
+        }
+
+        let asset_id = req.asset;
+        if !self.balance_manager.asset_manager.asset_exist(&asset_id) {
+            return Err(Status::invalid_argument("invalid asset"));
+        }
+
+        let from_user_id = req.from;
+        let to_user_id = req.to;
+
+        let balance_manager = &self.balance_manager;
+        let balance_from = balance_manager.get(from_user_id, BalanceType::AVAILABLE, &asset_id);
+
+        let zero = Decimal::from(0);
+        let delta = Decimal::from_str(&req.delta).unwrap_or(zero);
+
+        if delta <= zero || delta > balance_from || !balance_manager.exists(to_user_id, BalanceType::AVAILABLE, &asset_id) {
+            return Ok(TransferResponse {
+                success: false,
+                asset: asset_id,
+                balance_from: balance_from.to_string(),
+            });
+        }
+
+        let prec = self.balance_manager.asset_manager.asset_prec_show(&asset_id);
+        let change = delta.round_dp(prec);
+
+        let business = "transfer".to_owned();
+        let business_id = Instant::now().elapsed().as_secs();
+
+        // Tries to update balances for both from and to users synchronously.
```

Comment:
We use https://github.com/Fluidex/dingir-exchange/blob/master/src/matchengine/server.rs#L13 so each grpc request will be handled synchronously, and global state can be consistent.

I admit that this still cannot guarantee atomicity. But your approach cannot guarantee atomicity either. 

But luckily we only use operation_log for finality. So if the server fails after decreasing user_from before increasing user_to, it won't really be recorded if the server restarts and replays the operation_log.

My thought will be it is not necessary to add a new function `transfer_user_balance`, but calling two `update_user_balance` (one for sub, one for add). We can also avoid redundant checks in both Controller::transfer and BalanceUpdateController::transfer_user_balance